xdsl/dialects/omp.py

```diff
@@ -583,6 +585,76 @@ def num_block_args(self) -> int:
         return len(self.private_vars) + len(self.reduction_vars)
 
 
+@irdl_op_definition
+class DeclareReductionOp(IRDLOperation):
+    """
+    Implementation of upstream omp.declare_reduction
+    See external [documentation](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/ODS/#ompdeclare_reduction-ompdeclarereductionop).
+    """
+
+    name = "omp.declare_reduction"
+
+    sym_name = prop_def(SymbolNameConstraint())
+    var_type = prop_def(TypeAttribute, prop_name="type")
+
+    alloc_region = region_def()
+    init_region = region_def()
```

Comment:
```suggestion
    initializer_region = region_def()
```
Nit: Would it make more sense to stick with upstream naming?